### ISSUE TITLE
Python 3 binding

### ISSUE THREAD

**Issue by cyang-kth** [cyang-kth/fmm#5]
Design a Python interface similar to the one in Caffe
https://github.com/BVLC/caffe/blob/master/python/caffe/_caffe.cpp

**Comment by zhuang-hao-ming** [comment]
Hello @cyang-kth .

I am not very familiar with Caffe.
I want to know what your definition of python is.
I just implemented a python interface.
It look like this.

```
>>> import fmm
>>> fmm.fmm("fmm_config.xml")
------------ Fast map matching (FMM) ------------
------------     Author: Can Yang    ------------
------------   Version: 2018.03.09   ------------
------------     Applicaton: fmm     ------------
Validating configuration for map match application:
    Warning, overwrite existing result file.mr.txt
```
And the code look like this.

```
#include <boost/python.hpp>
#include <string>
using namespace std;

#include <app/fmm.cpp>

void fmm(string config_file_name) {

	char* argvs[] = { "fmm" , const_cast<char*>(config_file_name.c_str()) };
	main(2, argvs);
}


BOOST_PYTHON_MODULE(fmm)
{
	using namespace boost::python;

	def("fmm", fmm);
}

```

I think there is no difference between such a python interface and the command line.
So I think you might have some suggestions on the design of python interfaces.

Thank.

**Comment by cyang-kth** [comment]
@zhuang-hao-ming 
This is a good question. What I expect for a python interface is something that can be tested interactively on the fly.  It would be convenient if different components of the program can be designed with Python interface. 

The program is like
```
from fmm import ubodt, network, mm_app
app = mm_app()
table = ubodt.read("ubodt.txt")
net = network.read("network.shp")
app.register_ubodt(table)
app.register_network(net)
app.register_config({r:200,k:8,pf:100})
# Some functions
print table.query_dist(12,23)
print table.query_path(12,23)
print network.search_candidate(point, k = 5, r= 200)
print app.match_wkt("Linestring ...")
```

With the python interface above, it would also be easy to design a map matching web service in Python using lightweight server such as flask. 

https://github.com/BVLC/caffe/tree/master/examples/web_demo

I have very few experience with writing Python interface for C program. Perhaps that would be quite a lot of work.

**Comment by cyang-kth** [comment]
There seems to be several ways to write python wrapper for c++.

- Boost python https://www.boost.org/doc/libs/1_69_0/libs/python/doc/html/index.html
- SWIG  http://www.swig.org/tutorial.html
- Cython https://cython.org/

### Other popular libraries 

- tensorflow : swig https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/pywrap_dlopen_global_flags.py#L31
- caffe: Cython
- GDAL: swig
- Numpy: swig

**Comment by cyang-kth** [comment]
A Python API has been added in the latest commit.

**Comment by cweber9843** [comment]
Thanks @cyang-kth for the great work with this code! 
Do you have plans to make the API available in python3? 
Would be great, especially since python 2.7 reached end of life in 2020.

**Comment by cyang-kth** [comment]
@cweber9843 

The current Python API is built using Swig, which supports Python 3.0 according to the documentation. 
http://www.swig.org/Doc3.0/SWIGDocumentation.html#Python_python3support

If you want to build the Python 3 API you can just add a line to this cmake file before this line
https://github.com/cyang-kth/fmm/blob/dafe91305781cbc4b12432b795a2428866076141/python/CMakeLists.txt#L68

```
set_property(TARGET fmm PROPERTY SWIG_COMPILE_OPTIONS -py3)
```
Then you should have the Python 3 binding installed. 

Currently I have no time to test the Python 3 binding but in the future it can be added as one option to the cmake command line to enable python 3 support.

**Comment by cweber9843** [comment]
Thanks for the quick reply! And thanks for the hint, I now managed to compile to python3.8, and the `fmm_test.py` completes successfully. Also my first test on my own data looks promising.

I had to do some more modifications in the `CMakeLists.txt`:
Require right python version:
`find_package(PythonLibs 3.8 REQUIRED)`
And I think the target is called `pyfmm`, not `fmm`:
`set_property(TARGET pyfmm PROPERTY SWIG_COMPILE_OPTIONS -py3)`
